torch/csrc/autograd/record_function.cpp

```diff
@@ -7,17 +7,18 @@ namespace {
 bool has_callbacks = false;
 std::vector<RecordFunctionCallback> start_callbacks;
 std::vector<RecordFunctionCallback> end_callbacks;
+std::vector<bool> needs_inputs_flags;
```

Comment:
why do you need both fields. You can just keep on your stack the "cumsum" of the inputs. I.e. `needs_inputs_flags.push_back(needs_inputs_flags.back() | needs_inputs);` (and push one neutral element to start)

Reply:
I didn't seem to me that this needed to be optimized considering readability and infrequency of adding new callbacks, but sure we could rewrite it with cumsum